### PR TITLE
Feature notifications

### PR DESCRIPTION
**Pull request by seblaifa3315**
### What this PR does (required):
I created Model, routes and controller for notifications

### Any issues with the current functionality (optional):
- I didn't completely understand the ticket... so I'm not sure if I did what was asked. please, let me know.

### CODE REVIEW

**Review by ezeilo-su — COMMENTED** [review]
Looks good!

File: server/models/Notification.js [hunk 1]
```diff
@@ -0,0 +1,41 @@
+const mongoose = require("mongoose");
+
+const notificationSchema = new mongoose.Schema(
```
Comment (rajivtitus) [inline server/models/Notification.js:3]: Nice model and routes 💯

File: server/models/Notification.js [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+            default: Date.now(),
+        },
+        senderId: {
+            type: mongoose.Schema.Types.ObjectId,
+            required: true,
+            ref: "User",
+        },
+        recipientId: {
+            type: mongoose.Schema.Types.ObjectId,
+            ref: "User",
```
Comment (rajivtitus) [inline server/models/Notification.js:33]:
```suggestion
            ref: "User",
            required: true,
```
I think a notification will always belong to a recipient :)

File: server/app.js [hunk 3]
```diff
@@ -47,6 +50,7 @@ app.use((req, res, next) => {
 app.use("/auth", authRouter);
 app.use("/users", userRouter);
 app.use("/profile", profileRouter);
+app.use("/notification", notificationRouter);
```
Comment (rajivtitus) [inline server/app.js:53]:
```suggestion
app.use("/notifications", notificationRouter);
```
Nit: Naming our routes as per REST guidelines

File: server/controllers/notification.js [hunk 4]
```diff
@@ -0,0 +1,60 @@
+const Notification = require("../models/Notification");
+const User = require("../models/User");
+const asyncHandler = require("express-async-handler");
+
+exports.createNotification = asyncHandler(async (req, res, next) => {
+    try {
```
Comment (rajivtitus) [inline server/controllers/notification.js:6]: We don't need a try/catch block as the `asyncHandler` will handle promise rejections and forward it to our error middleware.

Ditto where applicable!

File: server/controllers/notification.js [hunk 5]
```diff
@@ -0,0 +1,60 @@
...
+        await theNotification.save();
+        res.status(200).json(theNotification);
+    } catch (err) {
+        res.status(500).json({ message: error.message });
+    }
+});
+
+exports.getAll = asyncHandler(async (req, res, next) => {
+    try {
+        const {type, recipientId} = req.body;
```
Comment (rajivtitus) [inline server/controllers/notification.js:28]: It is recommended not to send any information within the body of a GET request

Once we've added the `protect` middleware to our routes, we can access the logged in user's id using the request object - `req.user.id`

File: server/controllers/notification.js [hunk 6]
```diff
@@ -0,0 +1,60 @@
...
+        
+        if(type !== "message"){
```
Comment (rajivtitus) [inline server/controllers/notification.js:30]: I don't understand the reason behind this check, I think we can straight away return all the notifications for a logged in user

File: server/controllers/notification.js [hunk 7]
```diff
@@ -0,0 +1,60 @@
...
+            const AllNotifications = await Notification.find();
+            if (AllNotifications.length === 0) res.status(404).json({ message: "There is no notification" });
+            res.status(200).json(AllNotifications);
+        } else {
+            const notifications = await Notification.find({ recipientId: recipientId });
+            if (notifications.length === 0) res.status(404).json({ message: "There is no notification" });
```
Comment (rajivtitus) [inline server/controllers/notification.js:36]:
```suggestion
```
A mongoose `find()` query returns an empty list if no records are matched. Ideally, we want to return this empty list to the FE to have a consistent response format

File: server/controllers/notification.js [hunk 8]
```diff
@@ -0,0 +1,60 @@
...
+            res.status(200).json(notifications);
+        }
+    } catch (err) {
+        res.status(500).json({ message: error.message });
+    }
+});
+
+exports.getUnread = asyncHandler(async (req, res, next) => {
```
Comment (rajivtitus) [inline server/controllers/notification.js:44]: Same comments as the controllers above can be applied here :)

**Review by rajivtitus — COMMENTED** [review]
Good work with this! Some comments for you to have a look and address 👍🏽

**Pull request merged**